#1959. Tests for syntax and basic restrictions

sgrekhov opened the pull request.

eernstg commented on LanguageFeatures/Class-modifiers/syntax_A06_t02.dart:
```diff
@@ -0,0 +1,27 @@
+
+class C {}
+
+mixin class MixinClass extends C {
```
Even `mixin class MixinClass extends Object {}` should probably be an error (it may or may not be a syntax error), because there is very little value in supporting that particular construct.

sgrekhov replied:
Ups... Spec:

> It's a compile-time error if a mixin class declaration:
> ...
> - has an extends clause,

It's an issue that `mixin class MixinClassExtendsObject extends Object {}` is not an error

eernstg replied:
Good catch! So let's test that case as well.

eernstg reviewed (commented):
Ah, I just discovered that the ordering of declarations is based on the ordering in the table in the class-modifiers feature specification (at least, it looks like that is the underlying principle). So we might very well just keep that ordering.

I did suggest a few other ways to order the declarations, based on their capabilities.

If you look at the table in the feature specification, we can see that the "No" and "Yes" entries in the table aren't a simple sequence of binary numbers (like 000, 001, 010, 011 etc., that is, No/No/No, No/No/Yes, No/Yes/No, No/Yes/Yes, etc.). It's arguably simpler to reason about the actual capabilities when the ordering is controlled directly by the capabilities (in order, or in the reverse order, that shouldn't matter much).

In the end it's probably fine to stick with the ordering in the feature spec. I guess the lists of declarations were created by looking at that ordering, and then picking the ones that would work for the given situation? (E.g., if we're testing a situation where the declarations must be `base`/`final`/`sealed` then we have to omit all the declarations where that is not true, because they are being tested in some other library where we expect errors.)

sgrekhov commented on LanguageFeatures/Class-modifiers/class_modifiers_lib.dart:
```diff
@@ -0,0 +1,27 @@
+base class BaseClass {}
+
+base mixin BaseMixin {}
+base mixin class BaseMixinClass {}
```
Oh yes! Of course! And `abstract base class`, `abstract final` etc... Thank you! I'll add these cases!

sgrekhov commented:
Yes, the ordering of the modifiers was taken from the table from the spec. It's easier just to follow that order from the spec, so I'd like to preserve it. Updating according to the pther notes now

eernstg commented:
> It's easier just to follow that order from the spec, so I'd like to preserve it

Very good!

sgrekhov commented on LanguageFeatures/Class-modifiers/basic_restrictions_A01_t03.dart:
```diff
@@ -0,0 +1,83 @@
+
+import "class_modifiers_lib.dart";
+
+mixin MixinOnSealed on SealedClass {}
```
Updated

sgrekhov commented on LanguageFeatures/Class-modifiers/basic_restrictions_A01_t04.dart:
```diff
@@ -0,0 +1,100 @@
+abstract base mixin class AbstractBaseMixinClass
+    implements ExtendsSealedClass {}
+
+mixin Mixin implements ExtendsSealedClass {}
```
Fixed

sgrekhov reviewed (commented):
Tests updated. Please rereview

eernstg reviewed (commented):
Just one thing missing: The treatment of `enum` declarations as having an invisible `final` modifier.

Taking another look at all the enum declarations, I think the following declarations should not be considered errors:

LanguageFeatures/Class-modifiers/basic_restrictions_A04_t01.dart: 144, 149
LanguageFeatures/Class-modifiers/basic_restrictions_A04_t02.dart: 141, 146, 151
LanguageFeatures/Class-modifiers/basic_restrictions_A04_t03.dart: 124
LanguageFeatures/Class-modifiers/basic_restrictions_A04_t04.dart: 120, 125
LanguageFeatures/Class-modifiers/basic_restrictions_A04_t06.dart: 79
LanguageFeatures/Class-modifiers/basic_restrictions_A04_t08.dart: 79

Please double check that they are only considered errors because of the (not so much) missing `base`/`final`/`sealed` modifier.

sgrekhov commented:
`base_restrictions_A04_t*` updated. 
- Removed test which checks that `implements BaseClass` where `BaseClass` is from another library is an error. It is tested in `base_restrictions_A03_t*`
- Added checks that `base/final/sealed class C implements BaseClass` is not an error if both `C` and `BaseClass` from the same library (`FinalClass` as well)
- enums are treated as `final` classes

Please rereview

eernstg reviewed (approved):
LGTM!
